airflow/sensors/smart_sensor_operator.py

```diff
@@ -0,0 +1,730 @@
+                self.log.exception("Exception at creating sensor work for ti %s", ti.key)
+                self.log.exception(e, exc_info=True)
+
+        self.log.info("%d tasks detected.", len(sensor_works))
+
+        new_sensor_works = [x for x in sensor_works if x not in self.sensor_works]
+
+        self._update_ti_hostname(new_sensor_works)
+
+        self.sensor_works = sensor_works
+
+    @provide_session
+    def _update_ti_hostname(self, sensor_works, session=None):
+        """
+        Update task instance hostname for new sensor works.
+        :param sensor_works:
+        :return:
+        """
+        TI = TaskInstance
+        ti_keys = [(x.dag_id, x.task_id, x.execution_date) for x in sensor_works]
+
+        def update_ti_hostname_with_count(count, ti_keys):
+            # Using or_ instead of in_ here to prevent from full table scan.
+            tis = session.query(TI) \
+                .filter(or_(tuple_(TI.dag_id, TI.task_id, TI.execution_date) == ti_key
+                            for ti_key in ti_keys)) \
+                .all()
+
+            for ti in tis:
+                ti.hostname = self.hostname
+            session.commit()
+
+            return count + len(ti_keys)
+
+        count = helpers.reduce_in_chunks(update_ti_hostname_with_count, ti_keys, 0, self.max_tis_per_query)
+        if count:
+            self.log.info("Updated hostname on %s tis.", count)
+
+    @provide_session
+    def _mark_multi_state(self, operator, poke_hash, encoded_poke_context, state, session=None):
+        """
+        Mark state for multiple tasks that have hashcode=poke_hash.
+        :param poke_hash:
+        :param state:
+        :param session:
+        :return:
+        """
+
+        def mark_state(ti, sensor_instance):
+            ti.state = state
+            sensor_instance.state = state
+            if state in State.finished():
+                ti.end_date = end_date
+                ti.set_duration()
+
+        SI = SensorInstance
+        TI = TaskInstance
+
+        count_marked = 0
+        try:
+            query_result = session.query(TI, SI)\
+                .join(TI, and_(TI.dag_id == SI.dag_id,
+                               TI.task_id == SI.task_id,
+                               TI.execution_date == SI.execution_date)) \
+                .filter(SI.state == State.SENSING) \
+                .filter(SI.hashcode == poke_hash) \
+                .filter(SI.operator == operator) \
+                .with_for_update().all()
+
+            end_date = timezone.utcnow()
+            for ti, sensor_instance in query_result:
+                if sensor_instance.poke_context != encoded_poke_context:
+                    continue
+
+                ti.hostname = self.hostname
+                if ti.state == State.SENSING:
+                    mark_state(ti=ti, sensor_instance=sensor_instance)
+                    count_marked += 1
+                else:
+                    # ti.state != State.SENSING
```

Comment:
Can this be removed?

Reply:
Task instance states could be changed by user activities such as marking dagrun and ti states from UI. The `sensor_instance` table will respect `task_instance` state changing that happened during it's poking phase. So if a user mark a ti state, it will not be overwritten by smart sensor.